File: src_java/AStar.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import datastructures.AStarSolution;
import datastructures.CLCSInstance;
import datastructures.Node;
import datastructures.NodePriorityQueue;
import datastructures.Pv_Uv;

public class AStar {

    public static AStarSolution AStarRun(CLCSInstance inst) {
        AStarSolution solution = new AStarSolution();
        NodePriorityQueue Q_p = new NodePriorityQueue();

        Map<List<Integer>, List<Integer[]>> N_v = new HashMap<>();

        Node root = new Node(inst.getNumInputs());
        root.setPriority(Upperbound.ubBoth(inst, root));

        int expanded = 0;

        Q_p.add(root);
        insertNewNodeToNv(N_v, root);

        while (!Q_p.isEmpty()) {
            Node v = Q_p.poll();
            List<Integer> v_pv = v.getPv();

            // Get visited nodes for this position
            List<Integer[]> N_vRelative = N_v.get(v_pv);

            // Skip node if the position is visited and the node is outdated
            if (N_vRelative != null && isOutdated(v, N_v.get(v_pv)))
                continue;

            // Node has not been visited. Insert node to the N hash map
            insertNewNodeToNv(N_v, v);

            List<Pv_Uv> v_nd = getFeasibleNonDominatedExtensions(inst, v);
            expanded++;

            if (v_nd.isEmpty()) {
                List<Integer> sol = deriveSolution(v, inst.getInputs().get(0));
                solution.getSolutions().add(sol);
                solution.getExpandeds().add(expanded);
                break;
            }

            for (Pv_Uv pv_Uv : v_nd) {
                int vExtendedL = v.getL_v() + 1;
                int vExtendedU = pv_Uv.getU_v();
                List<Integer> vExtendedPv = pv_Uv.getPv();

                boolean doInsert = true;

                N_vRelative = N_v.get(vExtendedPv);
                boolean isVExtVisited = N_vRelative != null;

                if (isVExtVisited) {
                    List<Integer[]> toRemove = new ArrayList<>();
                    doInsert = isExtNotDominated(N_vRelative, vExtendedL, vExtendedU, toRemove);
                    for (Integer[] vRelative : toRemove) {
                        Q_p.remove(pv_Uv.getPv(), vRelative[0], vRelative[1]);
                        N_vRelative.remove(vRelative);
                    }
                }

                if (doInsert) {
                    Integer[] vExtendedLU = { vExtendedL, vExtendedU };
                    Node vExtended = new Node(inst.getNumInputs(), vExtendedPv, vExtendedL, vExtendedU, v);
                    vExtended.setPriority(Upperbound.ubBoth(inst, vExtended));
                    if (isVExtVisited) {
                        N_vRelative.add(vExtendedLU);
                    } else {
                        insertNewNodeToNv(N_v, vExtended);
                    }
                    Q_p.add(vExtended);
                }
            }
        }

        return solution;
    }

    private static boolean isExtNotDominated(List<Integer[]> N_vRelative, int vExtendedL, int vExtendedU, List<Integer[]> toRemove) {
        boolean doInsert = true;
        for (Integer[] vRelative : N_vRelative) {
            int vRelativeL = vRelative[0];
            int vRelativeU = vRelative[1];

            // A node v1 dominates v2 if and only if
            // lv1 >= lv2 && uv1 >= uv2

            boolean L_relativeEQ = vRelativeL == vExtendedL;
            boolean U_relativeEQ = vRelativeU == vExtendedU;

            // Condition 1: Can be eased out to find equally good sols
            boolean L_relativeChosen = (vRelativeL > vExtendedL) || L_relativeEQ;
            boolean U_relativeChosen = (vRelativeU > vExtendedU) || U_relativeEQ;

            // Condition 2: Without, can lead to very long run time
            boolean L_relativeReject = (vRelativeL < vExtendedL) || L_relativeEQ;
            boolean U_relativeReject = (vRelativeU < vExtendedU) || U_relativeEQ;

            // If v_ext is dominated by a known node v_rel,
            // we stop the comparison and skip v_ext.
            if (L_relativeChosen && U_relativeChosen) {
                doInsert = false;
                break;
            }

            if (L_relativeReject && U_relativeReject) {
                toRemove.add(vRelative);
            }
        }
        return doInsert;
    }

    private static boolean isOutdated(Node v, List<Integer[]> N_vRelative) {
        boolean outdated = false;
        // Iterates over each pair of partial solution length and
        // constraint prefix length in the list of pairs for the node's
        // position vector in the hash map.
        for (Integer[] v_rel : N_vRelative) {
            outdated = isOutdatedSingle(v, v_rel);
            if (outdated)
                break;
        }
        return outdated;
    }

    private static boolean isOutdatedSingle(Node v, Integer[] v_rel) {
        int l_rel = v_rel[0];
        int u_rel = v_rel[1];
        // If the partial solution length and constraint prefix length of the node
        // are smaller than those of a pair in the list, the node is outdated.
        boolean cond1 = l_rel > v.getL_v() && u_rel >= v.getU_v();
        boolean cond2 = l_rel == v.getL_v() && u_rel > v.getU_v();
        if (cond1 || cond2)
            return true;
        return false;
    }

    private static void insertNewNodeToNv(Map<List<Integer>, List<Integer[]>> N_v, Node v) {
        List<Integer[]> tmp_v_childs = new ArrayList<>();
        tmp_v_childs.add(v.getLU());
        N_v.put(v.getPv(), tmp_v_childs);
    }

    private static List<Pv_Uv> getFeasibleNonDominatedExtensions(CLCSInstance inst, Node v) {
        List<Pv_Uv> feasibles = new ArrayList<>();

        // Return empty if reached the end of any sequence (Cannot satisfy constraint if
        // go further)
        if (isOvershoot(inst.getNumInputs(), v.getPv(), inst.getInputLengths())) {
            return feasibles;
        }

        // Get non-dominated potential extensions
        List<Integer> sigmaND = getSigmaND(inst.getNumInputs(), inst.getSigmaLength(), v.getPv(),
                inst.getSuccessorTables());

        // If the partial solution already satisfy the constraint P,
        // then we don't need to check for feasibility

        if (v.getU_v() == inst.getConstraintLength()) {
            for (Integer label : sigmaND) {
                feasibles.add(
                        PvUvFromLabel(label, inst.getNumInputs(), v.getU_v(), v.getPv(), inst.getSuccessorTables()));
            }
            return feasibles;
        } 

        for (Integer label : sigmaND) {
            boolean isNextConstraint = (label == inst.getConstraint().get(v.getU_v()));

            // If the label is not the next missing P element 
            // and adding it to the partial solution would lead to a solution 
            // where the constraint P cannot be met 
            if(!isNextConstraint && !isFeasible(label, v.getU_v(), inst.getNumInputs(), v.getPv(), inst.getSuccessorTables(), inst.getEmbedTables())) {
                continue;
            } 
            // If the label matches the next missing element of P, increment u_v
            int newU_v = isNextConstraint ? v.getU_v() + 1 : v.getU_v();
            feasibles.add(
                        PvUvFromLabel(label, inst.getNumInputs(), newU_v, v.getPv(), inst.getSuccessorTables()));
        }

        return feasibles;
    }

    private static boolean isOvershoot(int numInputs, List<Integer> positionVector, int[] inputLengths) {
        // Iterates over each input sequence
        // If any sequence is overshot
        for (int i = 0; i < numInputs; i++) {
            if (positionVector.get(i) > inputLengths[i])
                return true;
        }
        return false;
    }

    private static List<Integer> getSigmaND(int numInputs, int sigmaLength, List<Integer> positionVector,
            int[][][] successorTables) {
        List<Integer> sigmaND = new ArrayList<>();

        // Iterates over each label in the alphabet
        for (int label = 0; label < sigmaLength; label++) {
            // If the label is not dominated, add it to the set of non-dominated labels
            if (!isDominated(label, numInputs, sigmaLength, positionVector, successorTables)) {
                sigmaND.add(label);
            }
        }
        return sigmaND;
    }

    private static boolean isDominated(int label, int numInputs, int sigmaLength, List<Integer> positionVector,
            int[][][] successorTables) {
        // If the label does not appear in the remainder of the input sequences,
        // it is dominated
        if (isLabelAbsent(label, numInputs, positionVector, successorTables)) {
            return true;
        }

        for (int labelOther = 0; labelOther < sigmaLength; labelOther++) {
            // Skip the potential extension itself
            if (label == labelOther)
                continue;

            // If the label appears later than another label in both strings, it is
            // dominated
            if (isDominatedByOther(label, labelOther, numInputs, positionVector, successorTables)) {
                return true;
            }
        }
        return false;
    }

    private static boolean isLabelAbsent(int label, int numInputs, List<Integer> positionVector,
            int[][][] successorTables) {
        for (int i = 0; i < numInputs; i++) {
            int currPos = positionVector.get(i) - 1;
            int labelSuccessorIDX = successorTables[i][label][currPos];
            if (labelSuccessorIDX == Integer.MAX_VALUE)
                return true;
        }
        return false;
    }

    private static boolean isDominatedByOther(int label, int labelOther, int numInputs, List<Integer> positionVector,
            int[][][] successorTables) {
        for (int i = 0; i < numInputs; i++) {
            int currPos = positionVector.get(i) - 1;
            int labelSuccessorIDX = successorTables[i][label][currPos];
            int otherSuccessorIDX = successorTables[i][labelOther][currPos];
            if (labelSuccessorIDX < otherSuccessorIDX)
                return false;
        }
        return true;
    }

    private static Pv_Uv PvUvFromLabel(int label, int numInputs, int u_v, List<Integer> positionVector,
            int[][][] successorTables) {
        List<Integer> newPV = new ArrayList<>(numInputs);

        for (int i = 0; i < numInputs; i++) {
            int currPos = positionVector.get(i) - 1;
            int labelSuccessorIDX = successorTables[i][label][currPos];
            // Calculates the new position vector.
            // +1 to go to the next character
            // +1 to get the 1-based index
            newPV.add(i, labelSuccessorIDX + 2);
        }

        return new Pv_Uv(newPV, u_v);
    }

    public static boolean isFeasible(int label, int u_v, int numInputs, List<Integer> positionVector, int[][][] successorTables, int[][] embedTables) {
        // Iterates over each input sequence
        for(int i = 0; i < numInputs; i++) {
            int currPos = positionVector.get(i) - 1;
            int labelSuccessorIDX = successorTables[i][label][currPos];
            // If the index exceeds the constraint limit, it is not a feasible extension
            if(labelSuccessorIDX > embedTables[i][u_v]) return false;
        }
        return true;
    }

    private static List<Integer> deriveSolution(Node v, List<Integer> firstInput) {
        List<Integer> solution = new ArrayList<>();
        while (v.getParent() != null) {
            solution.add(0, firstInput.get(v.getPv().get(0) - 2));
            v = v.getParent();
        }
        return solution;
    }



}

File: src_py/astar.py
from datastructure import (
    Node,
    Pv_Uv,
    AStar_Solution,
    CLCS,
    # MaxPriorityQueueOptimized,
    # MaxPriorityQueueOptimizedSecond,
    MaxPriorityQueueOptimizedThird,
)
from astar_utils import (
    create_pv_uv_from_label,
    feasible_is_overshoot,
    get_sigma_nd,
    is_feasible,
    is_node_dominated,
)
from upperbound import ub_both


def astar_run(inst: CLCS):
    astar_sol = AStar_Solution.create()
    # Q_p = MaxPriorityQueueOptimized()
    # Q_p = MaxPriorityQueueOptimizedSecond()
    Q_p = MaxPriorityQueueOptimizedThird()
    N_v: dict[tuple[int, ...], list[tuple[int, int]]] = dict()
    label_blacklist: set = set()
    Node.set_pv_length(inst.num_inputs)
    root = Node.create()

    expanded = 0

    Q_p.push(root, ub_both(inst, root))

    while not Q_p.is_empty():
        v: Node = Q_p.pop()
        N_v_relative: list[tuple[int, int]] | None = N_v.get(v.pv)

        if N_v_relative is not None and check_is_outdated(v, N_v_relative):
            continue

        insert_new_to_nv(N_v, v)

        v_nd = get_feasible_non_dominated_extensions(inst, v, label_blacklist)
        expanded += 1

        if not v_nd:
            sol = derive_solution(v, inst.inputs[0])
            astar_sol.solutions.append(sol)
            astar_sol.expandeds.append(expanded)
            # print(f"Expanded {expanded}")
            break
        for pv_uv in v_nd:
            v_ext_l = v.l_v + 1
            v_ext_u = pv_uv.u_v
            v_ext_pv = pv_uv.pv

            v_ext_pv_tuple = tuple(v_ext_pv)

            do_insert = True

            N_v_relative = N_v.get(v_ext_pv_tuple)
            is_vext_visited = N_v_relative is not None

            if is_vext_visited:
                # assert N_v_relative is not None
                to_remove: list[tuple[int, int]] = []
                do_insert = is_ext_not_dominated(
                    N_v_relative,  # type: ignore
                    v_ext_l,
                    v_ext_u,
                    to_remove,
                )

                for v_rel in to_remove:
                    # Q_p.remove_node(v_ext_pv_tuple, v_rel[0], v_rel[1])
                    Q_p.remove_node((v_ext_pv_tuple, v_rel[0], v_rel[1]))
                    N_v_relative.remove(v_rel)  # type: ignore

            if do_insert:
                v_ext_l_u = (v_ext_l, v_ext_u)
                v_ext_node = Node(v_ext_pv_tuple, v_ext_l, v_ext_u, v)
                v_ext_ub = ub_both(inst, v_ext_node)
                if is_vext_visited:
                    # assert N_v_relative is not None
                    N_v_relative.insert(0, v_ext_l_u)  # type: ignore
                else:
                    insert_new_to_nv(N_v, v_ext_node)
                Q_p.push(v_ext_node, v_ext_ub)

    return astar_sol


def insert_new_to_nv(
    N_v: dict[tuple[int, ...], list[tuple[int, int]]], v: Node
) -> None:
    N_v[v.get_pv()] = [v.get_l_u()]


def check_is_outdated(v: Node, N_v_relative: list[tuple[int, int]]) -> bool:
    outdated = False
    for v_rel in N_v_relative:
        outdated = is_node_dominated(v, v_rel)
        if outdated:
            break
    return outdated


def derive_solution(v: Node, first_input: list[int]) -> list[int]:
    solution: list[int] = []
    while v.parent is not None:
        solution.insert(0, first_input[v.pv[0] - 2])
        v = v.parent
    return solution


def get_feasible_non_dominated_extensions(
    inst: CLCS, v: Node, label_blacklist: set
) -> list[Pv_Uv]:
    if feasible_is_overshoot(inst.num_inputs, v.pv, inst.input_lens):
        return []

    sigma_nd = get_sigma_nd(
        inst.num_inputs, inst.sigma_len, v.pv, inst.successor_tables, label_blacklist
    )

    feasibles: list[Pv_Uv] = []
    if v.u_v == inst.constraint_len:
        feasibles = [
            create_pv_uv_from_label(
                label, inst.num_inputs, v.u_v, v.pv, inst.successor_tables
            )
            for label in sigma_nd
        ]
        return feasibles

    for label in sigma_nd:
        vector = (label, v.pv)
        if vector in label_blacklist:
            continue

        is_next_constraint = label == inst.constraint[v.u_v]

        if not is_next_constraint and not is_feasible(
            label,
            inst.num_inputs,
            v.u_v,
            v.pv,
            inst.successor_tables,
            inst.embed_tables,
        ):
            continue

        new_uv = v.u_v + 1 if is_next_constraint else v.u_v
        feasibles.append(
            create_pv_uv_from_label(
                label, inst.num_inputs, new_uv, v.pv, inst.successor_tables
            )
        )

    return feasibles


def is_ext_not_dominated(
    N_v_relative: list[tuple[int, int]],
    vExtendedL: int,
    vExtendedU: int,
    to_remove: list[tuple[int, int]],
):
    do_insert = True

    for vRelative in N_v_relative:
        vRelativeL = vRelative[0]
        vRelativeU = vRelative[1]

        L_relativeEQ = vRelativeL == vExtendedL
        U_relativeEQ = vRelativeU == vExtendedU

        L_relativeChosen = (vRelativeL > vExtendedL) or L_relativeEQ
        U_relativeChosen = (vRelativeU > vExtendedU) or U_relativeEQ

        L_relativeReject = (vRelativeL < vExtendedL) or L_relativeEQ
        U_relativeReject = (vRelativeU < vExtendedU) or U_relativeEQ

        if L_relativeChosen and U_relativeChosen:
            do_insert = False
            break
        if L_relativeReject and U_relativeReject:
            to_remove.append(vRelative)

    return do_insert
